### PR TITLE
Python codegen: big cleaning and paving the way towards transforms

### PR DESCRIPTION
**Pull request by abey79**
### What

**Note**: as usual, review commit-by-commit (the bulk of the changes are in the first commit). This time, reviewers might want to spend some time on the auto-generated code, though, as this is a major focus of this PR.

This PR has the following aims:
- Enforce and exploit the decision that a component must be a struct of one, and exactly one, field, which may be a datatype (in which case the component is known as "delegating", as its implementation delegates to the datatype) or a native type (e.g. float, int, etc.).
- Significantly polish the Python SDK, with the goal of a cleaner, fully typed user-facing API, and a streamlined implementation. Part of this is achieved using `attrs` (see below).
- Improve the `from_similar()` hook and introduce several others. The proposed mechanisms consist of storing the so-called (hand-coded) overrides in the `_overrides` sub-packages in each of the `archetypes`, `components`, and `datatypes` packages. These overrides are discovered by the code generator by "parsing" (well, "brute-force extracting symbols" from) `_overrides/__init__.py` [1]. The following hooks are currently implemented:
    - `xxx_native_to_pa_array()` (where `xxx` is the lower-cased name of the corresponding object): this is the primary way to flexibly convert native input into the Arrow array form of the component.
    - `xxx_yyy_converter` (`yyy` is a field name): optional converter function a given field of a native objects (e.g. `Color`). This enables the object to flexibly accept native input for the given field.
    - `xxx_as_array`: optional `__array__()` magic function for native objects (e.g. `Point2D`). This function enable the object to be interpreted as array data by Numpy, which can considerably simplify the implementation of, e.g., `xxx_native_to_pa_array` (`Point2D` is a good example).

This PR also introduces the following changes:
- So far, the archetypes and `log_any()` were pulled into the `rr` namespace. This PR now pulls all component- and datatype-related symbols in `rr.cmp` and `rr.dt`, respectively. This is intended as a temporary solution, to avoid name clashes with the legacy API [2].
- The [`attrs`](https://www.attrs.org/en/stable/) dependency is introduced as it allows much more flexibility than the standard library `dataclass`, in particular with the `converter` feature which I use extensively. Introducing new deps is always a tough call. In the case of `attrs`, it's most probably safe as this is one of the most common and well maintained package of the ecosystem (110+M dl/month).
- Native objects automatically get `__str__()`, `__int__()`, or `__float__()` functions when they contain a single field of the related type.
- The `Point2D` component now delegate to a new `Point2D` datatype, defined as a struct of two floats. This is done such as to avoid breaking compatibility with the current viewer. The long-term goal is to use the `Vec2D` datatype instead.
- The CI now uses Python 3.11 for linting (and linting only, i.e. not testing, etc.)

Addressing union types is a non-goal of this PR. This will be part of the next PR (transforms). As such `QuotedObject::from_union()` is most likely broken.

TODO/TBD:
- PyCharm's typechecker isn't happy with the Archetype, with is irritating. This could apparently be addressed by using `xxxArrayLike` instead of `xxxArray` annotations for archetype's fields. I'm curious how VSCode fares, though.
- Maybe passing `None` as args to the archetype should set `None` instead of empty arrays (they aren't logged either way).
- Maybe the notion of delegating vs. non-delegating component should be promoted to the langauge-agnostic code?

[1] I actually prototyped and tested an approach based on PyO3. It sorta worked, until I tried to import stuff (Numpy, ...) for the overrides. The codegen would somehow be given access to a venv with the required deps, etc. Hell no.
[2] Name clashes also exist internally to the new API, e.g. `components.Point2D` vs. `datatypes.Point2D`. This will have to be addressed before we can pull "everything" into the `rr` namespace.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/2603) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/2603)
- [Docs preview](https://rerun.io/preview/pr%3Aantoine%2Fhope-python-cleanup-new/docs)
- [Examples preview](https://rerun.io/preview/pr%3Aantoine%2Fhope-python-cleanup-new/examples)

### CODE REVIEW

**Comment by teh-cmc** [comment]
Yeah my vim isn't too happy either... :(

image

File: crates/re_types_builder/src/codegen/python.rs [hunk 1]
```diff
@@ -16,6 +16,52 @@ use crate::{
...
+        }
+
+        let field = &self.fields[0];
+        if let Type::Object(name) = &field.typ {
+            return Some(objects.get(name));
+        }
+
+        unreachable!()
+    }
+}
```
Comment (teh-cmc) [inline crates/re_types_builder/src/codegen/python.rs:63]: That's some very Pythonic code :D

```suggestion
impl PythonObjectExt for Object {
    fn is_delegating_component(&self) -> bool {
        self.kind == ObjectKind::Component && matches!(self.fields[0].typ, Type::Object(_))
    }

    fn is_non_delegating_component(&self) -> bool {
        self.kind == ObjectKind::Component && !self.is_delegating_component()
    }

    fn delegate_datatype<'a>(&self, objects: &'a Objects) -> Option<&'a Object> {
        self.is_delegating_component()
            .then(|| {
                if let Type::Object(name) = &self.fields[0].typ {
                    Some(objects.get(name))
                } else {
                    None
                }
            })
            .flatten()
    }
}
```

File: crates/re_types_builder/src/codegen/python.rs [hunk 2]
```diff
@@ -28,18 +74,36 @@ impl PythonCodeGenerator {
     }
 }
 
+/// Inspect `_overrides` sub-packages for manual override of the generated code.
+///
+/// This is the hacky way. We extract all identifiers from `__init__.py` which contains, but don't
+/// start with, a underscore (`_`).
+fn load_overrides(path_buf: &Path) -> HashSet<String> {
+    let contents = std::fs::read_to_string(path_buf.join("_overrides").join("__init__.py"))
+        .expect("_overrides/__init__.py must exist");
```
Comment (teh-cmc) [inline crates/re_types_builder/src/codegen/python.rs:83]: The user would have no idea which of the overrides failed.

```suggestion
fn load_overrides(path: &Path) -> HashSet<String> {
    let path = path.join("_overrides").join("__init__.py");
    let contents = std::fs::read_to_string(&path)
        .with_context(|| format!("couldn't load overrides module at {path:?}"))
        .unwrap();
```

File: crates/re_types/definitions/rerun/testing/components/fuzzy.fbs [hunk 3]
```diff
@@ -56,17 +56,17 @@ table AffixFuzzer6 (
...
-  single_float_optional: float (order: 101);
-  single_string_required: string (order: 102, required);
-  single_string_optional: string (order: 103);
-  many_floats_optional: [float] (order: 104);
-  many_strings_required: [string] (order: 105, required);
-  many_strings_optional: [string] (order: 106);
-  // TODO(cmc): the ugly bug we need to take care of at some point
-  // many_transparent_optionals: rerun.testing.datatypes.AffixFuzzer2 (order: 107);
-}
+// table AffixFuzzer7 (
```
Comment (emilk) [inline crates/re_types/definitions/rerun/testing/components/fuzzy.fbs:59]: why is this block commented out?

Comment (teh-cmc) [reply crates/re_types/definitions/rerun/testing/components/fuzzy.fbs:59]: :point_up: as my review will point out, it shouldn't :grimacing:

**Review by teh-cmc — CHANGES_REQUESTED** [review]
Niceeee! Need to bring those Fuzzy things back tho!

**Comment by abey79** [comment]
> Yeah my vim isn't too happy either... :(
> 
> image

The situation here is a bit weird typing-wise TBH (due to the args being extracted from the big arrays, etc.). How does it look in `api_demo/main.py` (whatever the new name is)?

File: rerun_py/tests/unit/test_points2d.py
[the remark this answers is not in the document]
Comment (abey79) [reply rerun_py/tests/unit/test_points2d.py:1]: Matter of taste really—I'm fine either way. This was a step towards enabling using everything from a single `import` (I dislike libs which require 10 imports as soon as you use them), thus the short names. It's temporary though, so it doesn't matter much.

**Comment by teh-cmc** [comment]
Here's what I get for the [roundtrip test](https://github.com/rerun-io/rerun/blob/ea9d971460132863846d67bb1f74e5a2da2cf647/tests/python/roundtrips/points2d/main.py):
image

Here's `test_api`:
image

File: rerun_py/tests/unit/test_points2d.py [hunk 4]
```diff
@@ -4,11 +4,10 @@
 
 import numpy as np
 import rerun as rr
-from rerun._rerun2 import components as rrc
 
 # NOTE: uncomment these to get a better auto-completion experience...
 # from rerun_sdk import rerun2 as rr
-# from rerun_sdk.rerun2 import components as rrc
+# from rerun_sdk.rerun2 import components as rr.ext
```
[the remark this answers is not in the document]
Comment (abey79) [reply rerun_py/tests/unit/test_points2d.py:10]: 👍🏻

File: crates/re_types_builder/src/codegen/python.rs (hunk 2 above)
Comment (abey79) [reply crates/re_types_builder/src/codegen/python.rs:74]: TIL with_context

File: rerun_py/tests/unit/test_points2d.py
[the remark this answers is not in the document]
Comment (abey79) [reply rerun_py/tests/unit/test_points2d.py:1]: reverted

**Review by teh-cmc — APPROVED** [review]

**Pull request merged**